src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferAsSpanOverSubstring.cs

```diff
@@ -0,0 +1,312 @@
+        private static readonly LocalizableString s_localizableDescription = CreateResource(nameof(Resx.PreferAsSpanOverSubstringDescription));
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!RequiredSymbols.TryGetSymbols(context.Compilation, out RequiredSymbols symbols))
+                return;
+
+            context.RegisterOperationBlockStartAction(OnOperationBlockStart);
+
+            void OnOperationBlockStart(OperationBlockStartAnalysisContext context)
+            {
+                var invocations = PooledConcurrentSet<IInvocationOperation>.GetInstance();
+
+                context.RegisterOperationAction(context =>
+                {
+                    var argument = (IArgumentOperation)context.Operation;
+                    if (!symbols.IsAnySubstringInvocation(WalkDownImplicitConversions(argument.Value)))
+                        return;
+                    if (argument.Parent is not IInvocationOperation invocation)
+                        return;
+                    invocations.Add(invocation);
+                }, OperationKind.Argument);
+
+                context.RegisterOperationBlockEndAction(context =>
+                {
+                    foreach (var invocation in invocations)
+                    {
+                        //  We search for an overload of the invoked member whose signature matches the signature of
+                        //  the invoked member, except with ReadOnlySpan<char> substituted in for some of the 
+                        //  arguments that are Substring invocations.
+                        if (!GetBestSpanBasedOverloads(symbols, invocation, context.CancellationToken).IsEmpty)
+                        {
+                            Diagnostic diagnostic = invocation.CreateDiagnostic(Rule);
+                            context.ReportDiagnostic(diagnostic);
+                        }
+                    }
+                    invocations.Free(context.CancellationToken);
+                });
+            }
+        }
+
+        internal static IOperation WalkDownImplicitConversions(IOperation operation)
```

Comment:
I am wondering if we should update the `WalkDownConversions` helper to have an optional predicate function `Func<IConversionOperation, bool>`.

Reply:
I think that would be a good idea. I've wanted that function on several other analyzers. While we're talking about upgrading helpers, I've also wanted a `GetFirstOrDefaultWithParamater` overload that takes `ITypeSymbol`s instead of `ParameterInfo`s.